narwhal/primary/src/state_handler.rs

```diff
@@ -185,16 +255,195 @@ impl RandomnessState {
                 &mut rand::thread_rng(),
             ) {
                 Ok(output) => {
+                    if let Err(e) = self.vss_key_output.set(output.vss_pk.clone()) {
+                        error!("random beacon: unable to write VSS key to output: {e:?}")
+                    }
+                    let num_shares = output.shares.as_ref().map_or(0, |shares| shares.len());
                     self.dkg_output = Some(output);
+                    info!("random beacon: DKG complete with {num_shares} shares for this node");
+                }
+                Err(fastcrypto::error::FastCryptoError::NotEnoughInputs) => (), // wait for more input
+                Err(e) => error!("random beacon: error while processing DKG Confirmations: {e:?}"),
+            }
+            // Begin randomness generation.
+            if self.dkg_output.is_some() {
+                info!("random beacon: start randomness generation");
+                self.send_partial_signatures().await;
+            }
+        }
+    }
+
+    async fn update_narwhal_round(&mut self, round: Round) {
+        self.narhwal_round = round;
+        let next_leader_narwhal_round = round + (round % 2);
```

Comment:
From what I'm seeing, we don't have to compute `next_leader_narwhal_round`. Just checking `self.last_narwhal_round_sent < round` should have the same result. It may potentially make more sense too.